microservices performance comparison (#2632)
OpenLiberty/blogs: posts/2022-10-06-memory-footprint-throughput-update.adoc

seo-title: How does Open Liberty’s performance compare to other cloud-native Java runtimes - OpenLiberty.io
seo-description: While cold startup time is critically important for cloud functions (and Open Liberty's time-to-first-response is approximately 1 second, so it's no slouch), memory footprint and throughput are far more important for microservices architectures in which each running instance is likely to serve thousands of requests before it's replaced.
blog_description: "While cold startup time is critically important for cloud functions (and Open Liberty's time-to-first-response is approximately 1 second, so it's no slouch), memory footprint and throughput are far more important for microservices architectures in which each running instance is likely to serve thousands of requests before it's replaced."
open-graph-image: https://openliberty.io/img/blog/perfblog3a.png
additional_authors:
- name: Tim Pickett
  github: https://github.com/tpickett1
  image: https://avatars0.githubusercontent.com/tpickett1
---
= How does Open Liberty’s performance compare to other cloud-native Java runtimes
Graham Charters <https://github.com/gcharters>
:imagesdir: /
:url-prefix:
:url-about: /

> For many years, the most important metric for application servers was throughput: how many transactions or requests can we squeeze out of this application?  The adoption of finer-grained architecture styles, such as microservices and functions, has led to a broadening in the performance characteristics we should care about, such as _memory footprint_ and _throughput_.

Reviewer: this reads a bit strangely to me- because throughput is already mentioned in the first sentence it seems repetitious here. Unless the emphasis is to say "memory footprint _and_ throughput." A couple suggestions:

> For many years, the most important metric for application servers was throughput: how many transactions or requests can we squeeze out of this application?  The adoption of finer-grained architecture styles, such as microservices and functions, has led to a broadening in the performance characteristics we should care about, such as _memory footprint_ and _throughput_.

Reviewer: ```suggestion
For many years, the most important metric for application servers was throughput: how many transactions or requests can we squeeze out of this application?  The adoption of finer-grained architecture styles, such as microservices and functions, has broadened the performance characteristics we should care about, such as memory footprint _and_ throughput.
```

> For many years, the most important metric for application servers was throughput: how many transactions or requests can we squeeze out of this application?  The adoption of finer-grained architecture styles, such as microservices and functions, has led to a broadening in the performance characteristics we should care about, such as _memory footprint_ and _throughput_.

Reviewer: ```suggestion
For many years, the most important metric for application servers was throughput: how many transactions or requests can we squeeze out of this application?  The adoption of finer-grained architecture styles, such as microservices and functions, broadens the desirable performance characteristics to include _memory footprint_ as well as  throughput.
```

Author: I'm fine with these suggested changes

> Where you once deployed a small number of runtime instances for your server and application, now you may be deploying tens or hundreds of microservices or functions.  Each of those instances comes with its own server, even if it's embedded in a runnable jar or pulled in as a container image there's a server in there somewhere. While cold startup time is critically important for cloud functions (and link:{url-prefix}/blog/2019/10/30/faster-startup-open-liberty.html[Open Liberty's time-to-first-response is approximately 1 second], so it's no slouch), memory footprint and throughput are far more important for microservices architectures in which each running instance is likely to serve thousands of requests before it's replaced.

Reviewer: ```suggestion
Where you once deployed a small number of runtime instances for your server and application, now you might deploy tens or hundreds of microservices or functions.  Each of those instances comes with its own server. Even if it's embedded in a runnable jar or pulled in as a container image, there's a server in there somewhere. While cold start time is critically important for cloud functions (and link:{url-prefix}/blog/2019/10/30/faster-startup-open-liberty.html[Open Liberty's time-to-first-response is approximately 1 second], so it's no slouch), memory footprint and throughput are far more important for microservices architectures in which each running instance is likely to serve thousands of requests before it's replaced.
```

Author: ```suggestion
Where you once deployed a small number of runtime instances for your server and application, now you might deploy tens or hundreds of microservices or functions.  Each of those instances comes with its own server.  Even if it's embedded in a runnable jar or pulled in as a container image there's a server in there somewhere. While cold startup time is critically important for cloud functions (and link:{url-prefix}/blog/2019/10/30/faster-startup-open-liberty.html[Open Liberty's time-to-first-response is approximately 1 second], so it's no slouch), memory footprint and throughput are far more important for microservices architectures in which each running instance is likely to serve thousands of requests before it's replaced.
```

Author: Other changes are fine but I think we should keep the reference to 'startup time' to be called startup rather than start since linked blog post has 'startup' in multiple places.  The term 'startup' is regularly used in our WAS performance charts, presentations, etc.

If you've chosen Spring Boot for your application (yes, you can link:{url-prefix}/guides/spring-boot.html[use Spring Boot on Open Liberty]), then our measurements show an approximately 2x memory footprint benefit from link:https://developer.ibm.com/technologies/java/articles/modernize-and-optimize-spring-boot-applications/[running on Open Liberty] rather than Tomcat.  For example, the figure below shows the relative memory usage when running the link:https://github.com/spring-projects/spring-petclinic[Spring Boot Petclinic] application under load with a 4Gb heap:

[.img_border_light]
image::img/blog/perfblog2a.png[Memory footprint during load of using Spring Boot Petclinic application in Docker,width=70%,align="center"]

== Throughput - higher throughput equals cheaper infrastructure and license costs

The association between throughput and costs is simple: being able to put more work through a runtime means you can deploy smaller or fewer instances to satisfy demand. You’ll pay less on infrastructure and less on license costs, which are typically based on virtual CPU usage.

> Open Liberty also has significant throughput benefits when compared to other runtimes.  The figure below shows throughput measurements for the link:https://github.com/blueperf[Acme Air Microservices] benchmark.  We can see Open Liberty performs better than WildFly and significantly better than the other three runtimes:

Reviewer: I think this differentiation applied to the old graph from 2020- where tomee and helidon were significantly lower. Here TomEE and JBoss are  on par with wildfly

Author: ```suggestion
Open Liberty also has significant throughput benefits when compared to other runtimes.  The figure below shows throughput measurements for the link:https://github.com/blueperf[Acme Air Microservices] benchmark.  We can see Open Liberty performs better than WildFly and the other three runtimes:
```

[.img_border_light]
image::img/blog/perfblog3a.png[Throughput using Acme Air Microservices application in Docker,width=70%,align="center"]

> When we compare Spring Boot on Open Liberty throughput with Spring Boot on Tomcat, we can see from the figure below that there is an almost 2x throughput benefit with Open Liberty. This is a similar benefit to that shown in the previous TomEE measurement, which suggests that Open Liberty inherently has a ~2x higher throughput than Tomcat-based runtimes:

Reviewer: ```suggestion
When we compare Spring Boot on Open Liberty throughput with Spring Boot on Tomcat, we can see from the figure below that there is an almost 2x throughput benefit with Open Liberty. This benefit is similar to that shown in the previous TomEE measurement, which suggests that Open Liberty inherently has a ~2x higher throughput than Tomcat-based runtimes:
```

[.img_border_light]
image::img/blog/perfblog4a.png[Throughput using Spring Boot Petclinic application in Docker,width=70%,align="center"]

== Bringing it together

> In the previous sections we spoke about the importance of memory and throughput metrics for saving you money on your microservices deployments.  We saw individual measurements for each metric, but to get a picture of the overall benefit it's important to combine the two.  To do this, we can simply multiply the two benefits, the results of which are shown in the table below.  Of course, your results may vary and we'd recommend trying it out for yourself.  However, in our measurements, it’s not unrealistic to be able to run your workloads with, at most, half of the instances you’d need for other full Java runtimes:

Reviewer: ```suggestion
In the previous sections, we spoke about the importance of memory and throughput metrics for saving you money on your microservices deployments.  We saw individual measurements for each metric, but to get a picture of the overall benefit, it's important to combine the two.  To do this, we can simply multiply the two benefits, the results of which are shown in the table below.  Of course, your results may vary and we'd recommend trying it out for yourself.  However, in our measurements, it’s not unrealistic to be able to run your workloads with, at most, half of the instances you’d need for other full Java runtimes:
```
